File: Sample.java
// Time Complexity :
// Space Complexity :
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

//Find Disappered Numbers
//Time Complexity -> O(2n) -> O(n)
//Space Complexity -> O(n) -> space for hashset
class Solution {
    //Using hashSet
    public List<Integer> findDisappearedNumbers(int[] nums) {
        int n = nums.length;
        List<Integer> result = new ArrayList();
        Set<Integer> set = new HashSet();
        for(int i=0;i<n;i++){
            set.add(i+1);
        }
        for(int i=0;i<n;i++){
            if(set.contains(nums[i])){
                set.remove(nums[i]);
            }
        }

        for(int num : set){
            result.add(num);
        }

        return result;
    }
}

//Time Complexity -> O(n) 
//Space Complexity -> O(1) -> In place
class Solution {
    //In place
    public List<Integer> findDisappearedNumbers(int[] nums) {
        List<Integer> result = new ArrayList();
        int n = nums.length;
        for(int i=0;i<n;i++){
            int idx = Math.abs(nums[i]) -1;
            if(nums[idx]>0){
                nums[idx] = nums[idx]*(-1);
            }
        }
        for(int i=0;i<n;i++){
            if(nums[i]>0){
                result.add(i+1);
            }
            else{
                nums[i] = nums[i]*(-1);
            }
        }

        return result;
    }
}

//Game of Life
//Time Complexity ->O(8*m*n) -> O(m*n) because for each element in the matrix we have to check the 8 boundries
//Space Complexity-> O(1) -> as everything in place
class Solution {
    int[][] dir;
    public void gameOfLife(int[][] board) {
         // o->1 = shown by 2
         // 1 -> 0 = shown by 3
        int nRows = board.length;
        int nCols = board[0].length;
                           //right //down. //left. //up.    //upleft //downright.  //upright. //downleft
        dir = new int[][]{{0,1},   {1,0}, {-1,0},  {0,-1} , {-1,-1} ,{1,1} ,      {1,-1} ,   {-1,1}};

        for(int i = 0;i<nRows;i++){
            for(int j = 0;j<nCols;j++){
                int val = board[i][j];
                int count = activeCounts(board,i,j,nRows,nCols);
                if(val == 1 && (count <2 || count > 3)){
                    board[i][j] = 3;
                }
                if(val == 0 && count ==3){
                    board[i][j] = 2;
                }

            }
        }
        for(int i = 0;i<nRows;i++){
            for(int j = 0;j<nCols;j++){
                int val = board[i][j];
                if(board[i][j] == 3){
                    board[i][j] = 0;
                }
                if(board[i][j] == 2){
                    board[i][j] = 1;
                }

            }
        }
        
    }

    private int activeCounts(int[][] board,int i,int j,int nRows,int nCols ){
        int count = 0;
        int r =0,c =0;
        for(int[] dir : dir){
            r = i + dir[0];
            c = j + dir[1];
            if(r >=0 && c >= 0 && r< nRows && c < nCols && (board[r][c] == 1 || board[r][c] == 3)){
            count ++;
        } 
        }

        
        
        return count;
    }
}

File: Sample.py
# Problem 1: Find disappeared numbers in a given array having values from 1 to n (not sorted)
# Time Complexicity: O(n), Space Complexicity: O(1)

class Solution:
    
	"""
	Make use of indices to reverse sign of the numbers
	- at the end, the indices of positive numbers + 1 should return list of missing numbers
	"""
	
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        
        result = []; n = len(nums)
        if n == 0: return result
        else:
            for i in range(n):
                indx = abs(nums[i]) - 1
                if nums[indx] > 0 : nums[indx] *= -1
            for i in range(n):
                if nums[i] > 0:
                    result.append(i+1)
                else: nums[i] *= -1
        return result
		
		
# Problem 2: Game of life, make an element in a given array given by 0-dead, 1-alive using 4 rules
# Time Complexicity: O(m*n), Space Complexicity: O(1)

class Solution:		
  
    def gameOfLife(self, board: List[List[int]]) -> None:
        
        """
        Do not return anything, modify board in-place instead.
		- develop another function to count the neighbors (alive in original)
		- replace existing live by 1 and dead by 0
        """
        
        if len(board) == 0: return
        # Alive to Dead 1 -> 0 (-1)
        # Dead to Alive 0 -> 1 (+2)
        m, n = len(board), len(board[0])
        
        def count(i, j):
            dirs = [(-1,1), (0,1), (1,1), (-1,0), (1,0), (-1,-1), (0,-1), (1,-1)]; count = 0
            for dir_ in dirs:
                r = dir_[0] + i; c = dir_[1] + j
                if r >= 0 and c >= 0 and r < m and c < n and (board[r][c] == -1 or board[r][c] == 1):
                    count += 1
            return count

        for i in range(m):
            for j in range(n):
                # underpopulation and alive -> dead
                if count(i, j) < 2 and board[i][j] == 1:
                    board[i][j] = -1
                # overpopulated and alive -> dead
                if count(i, j) > 3 and board[i][j] == 1:
                    board[i][j] = -1
                # just right population
                if count(i, j) == 2 or count(i, j) == 3:
                    board[i][j] = board[i][j]
                # exactly 3 neighbors
                if count(i, j) == 3 and board[i][j] == 0:
                    board[i][j] = 2
                    
        for i in range(m):
            for j in range(n):
                if board[i][j] == -1:
                    board[i][j] = 0
                if board[i][j] == 2:
                    board[i][j] = 1
                    
        return board


# Extra question: find maximum and minimum element in a given array: (discussed in class), Time Complexicity: O(3*n/2) 
# {specifically mentioned in class, please verify}

def findMax(array):
    
    if len(array) == 0: return
    left = 0; right = len(array)-1

    while left < right:
        if array[left] < array[right]:
            left += 1
        else:
            right -= 1

    return array[left]

def findMin(array):
    
    if len(array) == 0: return
    left = 0; right = len(array)-1
    
    while left < right:
        if array[right] < array[left]:
            left += 1
        else:
            right -= 1

    return array[right]
